File: src/java/org/orbeon/oxf/xml/HTMLFragmentSerializer.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xml;

import org.xml.sax.Attributes;
import org.xml.sax.ContentHandler;
import org.xml.sax.Locator;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.io.Writer;

/**
 * Minimalist HTML fragment serializer.
 * 
 * o This assumes that all input elements are HTML or XHTML elements.
 * o Only attributes in no namespace are serialized.
 */
public class HTMLFragmentSerializer implements ContentHandler {

    private final Writer writer;
    private final boolean skipRootElement;

    private int level = 0;

    public HTMLFragmentSerializer(Writer writer, boolean skipRootElement) {
        this.writer = writer;
        this.skipRootElement = skipRootElement;
    }

    public void startDocument() {}
    public void endDocument() {}
    public void startPrefixMapping(String s, String s1) {}
    public void endPrefixMapping(String s) {}
    public void setDocumentLocator(Locator locator) {}

    public void startElement(String uri, String localname, String qName, Attributes attributes) throws SAXException {
        if (!skipRootElement || level > 0) {
            try {
                writer.write('<');
                writer.write(localname);
                final int attributesCount = attributes.getLength();
                if (attributesCount > 0) {
                    for (int i = 0; i < attributesCount; i++) {
                        final String currentAttributeName = attributes.getLocalName(i);
                        final String currentAttributeValue = attributes.getValue(i);

                        // Only consider attributes in no namespace
                        if ("".equals(attributes.getURI(i))) {
                            writer.write(' ');
                            writer.write(currentAttributeName);
                            writer.write("=\"");
                            if (currentAttributeValue != null)
                                writer.write(XMLUtils.escapeXMLMinimal(currentAttributeValue));
                            writer.write('"');
                        }
                    }
                }
                writer.write('>');
            } catch (IOException e) {
                throw new SAXException(e);
            }
        }
        level++;
    }

    public void endElement(String uri, String localname, String qName) throws SAXException {
        level--;
        if (!skipRootElement || level > 0) {
            try {
                writer.write("</");
                writer.write(localname);
                writer.write('>');
            } catch (IOException e) {
                throw new SAXException(e);
            }
        }
    }

    public void characters(char[] chars, int start, int length) throws SAXException {
        if (!skipRootElement || level > 0) {
            try {
                writer.write(XMLUtils.escapeXMLMinimal(new String(chars, start, length)));
            } catch (IOException e) {
                throw new SAXException(e);
            }
        }
    }

    public void ignorableWhitespace(char[] chars, int start, int length) throws SAXException {}
    public void processingInstruction(String s, String s1) throws SAXException {}
    public void skippedEntity(String s) throws SAXException {}
}

File: src/main/scala/org/orbeon/oxf/xml/HTMLFragmentSerializer.scala
/**
 * Copyright (C) 2015 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xml

import java.io.Writer

import javax.xml.transform.stream.StreamResult
import org.orbeon.dom.QName
import org.orbeon.io.CharsetNames
import org.orbeon.oxf.processor.XPLConstants
import org.orbeon.oxf.processor.converter.{TextConverterBase, XMLConverter}
import org.orbeon.oxf.properties.Properties
import org.xml.sax.Attributes

object HTMLFragmentSerializer {

  def create(writer: Writer, skipRootElement: Boolean): ForwardingXMLReceiver = {

    val identity = TransformerUtils.getIdentityTransformerHandler

    TransformerUtils.applyOutputProperties(
      identity.getTransformer,
      Properties.instance.getPropertySet(
        QName(
          "html-converter",
          XPLConstants.OXF_PROCESSORS_NAMESPACE
        )
      ).getQName(
        TextConverterBase.DEFAULT_METHOD_PROPERTY_NAME,
        XMLConverter.DEFAULT_METHOD
      ).clarkName,
      null,
      null,
      null,
      CharsetNames.Utf8,
      true,
      null,
      false,
      0
    )

    identity.setResult(new StreamResult(writer))

    val htmlReceiver = new PlainHTMLOrXHTMLReceiver("", identity)

    if (skipRootElement)
      new SkipRootElement(htmlReceiver)
    else
      htmlReceiver

  }

  class SkipRootElement(receiver: XMLReceiver) extends ForwardingXMLReceiver(receiver) {

    var level = 0

    override def startElement(uri: String, localname: String, qName: String, attributes: Attributes): Unit = {

      if (level > 0)
        super.startElement(uri, localname, qName, attributes)

      level += 1
    }

    override def endElement(uri: String, localname: String, qName: String): Unit = {

      level -= 1

      if (level > 0)
        super.endElement(uri, localname, qName)
    }
  }
}
